### PR TITLE
Refactor Chunks and ChildValues API to work iteratively

### PR DESCRIPTION
**Pull request by zcstarr**
Change exposes WalkRefs and WalkValues which has optional shallow or deep
traversal. This change also removes broken concurrency feature and dead 
code from the walk package

### CODE REVIEW

**Comment by zcstarr** [comment]
@rafael-atticlabs  PTAL wasn't sure if it would be best that I merged upstream changes prior to the PR or to do the PR after merging the upstream changes, wanted it to be fairly easy to track whats going on.

**Comment by zcstarr** [comment]
@rafael-atticlabs  PTAL merged changes from upstream back in

**Comment by rafael-atticlabs** [comment]
mostly lg. back to you.

**Comment by zcstarr** [comment]
@rafael-atticlabs  missing previous comments. That said I adjusted the code according to your comments IRL. PTAL

File: go/walk/walk.go [hunk 1]
```diff
@@ -11,41 +11,13 @@ import (
...
-		}
-		return false
-	}
-	doTreeWalkP(target, vr, callback, deep)
+	doTreeWalkP(target, vr, cb, deep)
 	return
 }
 
-func doTreeWalkP(v types.Value, vr types.ValueReader, cb SomeCallback, deep bool) {
+func doTreeWalkP(v types.Value, vr types.ValueReader, cb types.ValueCallback, deep bool) {
```
Comment (rafael-atticlabs) [inline go/walk/walk.go]: nit: `doTreeWalkP` => `doTreeWalk`

**Review by ghost — APPROVED** [review]
lgtm, w/ nit.

also note that we need to do this for js. @zcstarr, it's up to you if you want to take that on.

File: go/types/set_leaf_sequence.go [hunk 2]
```diff
@@ -24,11 +24,22 @@ func (sl setLeafSequence) getItem(idx int) sequenceItem {
 	return sl.data[idx]
 }
 
-func (sl setLeafSequence) Chunks() (chunks []Ref) {
+func (sl setLeafSequence) seqLen() int {
```
Comment (zcstarr) [inline go/types/set_leaf_sequence.go]: yep

**Comment by zcstarr** [comment]
@rafael-atticlabs  I ended up needing to change Walk API to support skipping, because photo-index demonstrated its utility. I also re-enabled the skip tests, which test skipping. PTAL

**Comment by rafael-atticlabs** [comment]
lgtm w/ comment fix & changes discussed in meatspace.

thanks for sticking with this! this is really nice now.

**Comment by zcstarr** [comment]
@rafael-atticlabs  thanks for the PR next stop js codebase!

**Pull request merged**